tor/controller.go

```diff
@@ -338,60 +338,79 @@ func (c *Controller) ProtocolInfo() ([]string, string, string, error) {
+type OnionType int
 
+const (
+	// V2 denotes that the onion service is V2.
+	V2 OnionType = iota
+
+	// V3 denotes that the onion service is V3.
+	V3
+)
+
+// AddOnionConfig houses all of the required paramaters in order to succesfully
+// create a new onion service or restore an existing one.
+type AddOnionConfig struct {
+	// Type denotes the type of the onion service that should be created.
+	Type OnionType
+
+	// VirtualPort is the externally reachable port of the onion address.
+	VirtualPort int
+
+	// TargetPorts is the set of ports that the service will be listening on
+	// locally. The virutal port will forward the traffic to one randomly
+	// chosen port from this set.
+	//
+	// NOTE: If nil/empty, the virtual port will be used as the only target
+	// port.
+	TargetPorts []int
+
+	// PrivateKeyPath is the full path to where the onion service's private
+	// key is stored. This can be used to restore an existing onion service.
+	PrivateKeyPath string
+}
+
+// AddOnion creates an onion service and returns its onion address. Once
+// created, the new onion service will remain active until the connection
+// between the controller and the Tor server is closed.
+func (c *Controller) AddOnion(cfg AddOnionConfig) (*OnionAddr, error) {
 	// We'll start off by checking if the file containing the private key
 	// exists. If it does not, then we should request the server to create
 	// a new onion service and return its private key. Otherwise, we'll
 	// request the server to recreate the onion server from our private key.
 	var keyParam string
-	if _, err := os.Stat(privateKeyFilename); os.IsNotExist(err) {
-		keyParam = "NEW:RSA1024"
+	if _, err := os.Stat(cfg.PrivateKeyPath); os.IsNotExist(err) {
+		switch cfg.Type {
+		case V2:
+			keyParam = "NEW:RSA1024"
+		case V3:
+			keyParam = "NEW:ED25519-V3"
+		}
 	} else {
-		privateKey, err := ioutil.ReadFile(privateKeyFilename)
+		privateKey, err := ioutil.ReadFile(cfg.PrivateKeyPath)
 		if err != nil {
 			return nil, err
 		}
 		keyParam = string(privateKey)
 	}
 
-	// Now, we'll determine the different virtual ports on which this onion
-	// service will be accessed by.
+	// Now, we'll create a mapping from the virtual port to each target
+	// port. If no target ports were specified, we'll use the virtual port
+	// to provide a one-to-one mapping.
 	var portParam string
-	for virtPort, targPorts := range virtToTargPorts {
-		// If the virtual port doesn't map to any target ports, we'll
-		// use the virtual port as the target port.
-		if len(targPorts) == 0 {
-			portParam += fmt.Sprintf("Port=%d,%d ", virtPort,
-				virtPort)
-			continue
-		}
-
-		// Otherwise, we'll create a mapping from the virtual port to
-		// each target port.
-		for targPort := range targPorts {
-			portParam += fmt.Sprintf("Port=%d,%d ", virtPort,
-				targPort)
+	if len(cfg.TargetPorts) == 0 {
+		portParam += fmt.Sprintf("Port=%d,%d ", cfg.VirtualPort,
+			cfg.VirtualPort)
+	} else {
```

Comment:
`else` not needed.

Reply:
I see how it's not needed, but IMO it helps with readability.